### PR TITLE
Reorganize the code base

### PR DESCRIPTION
**Pull request by lbittarello**
This PR breaks some files into smaller ones. It doesn't introduce any substantive changes.

Functions that are imported by another file have been renamed so they don't start with a leading underscore.

### CODE REVIEW

File: src/glum/_glm_regressor.py
Comment (stanmart) [inline src/glum/_glm_regressor.py:1]: Why not simply `_glm_py`, and the current `_glm.py` could become `_glm_base.py`. That would be a bit more aligned with class names IMO.

File: src/glum/_algebra.py
Comment (stanmart) [inline src/glum/_algebra.py:1]: What about `_linalg.py`? That would resemble numpy and scipy's submodules.

File: .github/workflows/build_wheels.yml [hunk 1]
```diff
@@ -52,7 +52,7 @@ jobs:
     if: github.event_name == 'release' && github.event.action == 'published'
     needs: [build_wheels, build_sdist]
     runs-on: ubuntu-latest
-    environment:
```
Comment (stanmart) [inline .github/workflows/build_wheels.yml:55]: Should we add the trailing whitespace fixer pre-commit-hook so that these don't reappear?

File: src/glum/_glm.py [hunk 2]
```diff
@@ -1,47 +1,19 @@
...
-import scipy.sparse.linalg as splinalg
 import sklearn as skl
 import tabmat as tm
-from formulaic import Formula, FormulaSpec
-from formulaic.parser import DefaultFormulaParser
-from formulaic.utils.constraints import LinearConstraintParser
-from formulaic.utils.context import capture_context
-from packaging import version
 from scipy import linalg, sparse, stats
-from sklearn.base import BaseEstimator, RegressorMixin
```
Comment (stanmart) [inline src/glum/_glm.py:37]: I think these imports are quite idiomatic, and that's how they appear in scikit-learn's docs, too.

File: tests/glm/test_glm_cv.py [hunk 3]
```diff
@@ -37,7 +36,7 @@ def test_normal_elastic_net_comparison(l1_ratio, fit_intercept, convert_x_fn):
     tol = 1e-9
 
     n_predict = 10
-    X, y, _ = make_regression(
+    X, y, _ = skl.datasets.make_regression(
```
Comment (stanmart) [inline tests/glm/test_glm_cv.py:39]: On the other hand I like using the qualified classname here as `make_regression` sounds like it could come from anywhere.

File: tests/glm/test_utils.py [hunk 4]
```diff
@@ -21,7 +21,7 @@ def df():
     )
 
 
-def test_align_df_categories_numeric(df):
+def testalign_df_categories_numeric(df):
```
Comment (stanmart) [inline tests/glm/test_utils.py:24]:
```suggestion
def test_align_df_categories_numeric(df):
```

And elsewhere. I imagine it's a typo.

File: tests/glm/test_utils.py [hunk 5]
```diff
@@ -180,7 +180,7 @@ def df_na():
     )
 
 
-def test_add_missing_categories(df_na):
+def testadd_missing_categories(df_na):
```
Comment (stanmart) [inline tests/glm/test_utils.py:183]:
```suggestion
def test_add_missing_categories(df_na):
```

**Review by stanmart — COMMENTED** [review]
Thank you, I think the new structure is much nicer than what we had before. `_glm.py` in particular was rather unwieldy. I just have a few small comments, plus the following bigger ones.

> Functions that are imported by another file have been renamed so they don't start with a leading underscore.

Where does that convention come from? I though we mostly use underscores to denote functions which are private _for the users of glum_. I understand that these functions live in underscored modules, which can mean the same thing, but why switch? I personally don't see a huge benefit in using names to denote whether functions are module-private or package-private. When working on glum, it's simple to use 'Go to references` and `Rename symbol` to make changes across files anyways.

This is not to say that I'm totally opposed to the proposed change. If the consensus is to follow the new convention, that's fine with me.

> This PR breaks some files into smaller ones. It doesn't introduce any substantive changes.

This is my more important point: we should still check downstream projects before merging this to make sure we don't break to much stuff the this refactor (or at least give them some time for them to prepare for it). I know that you are not changing the public API, but it would not surprise me if downstream users were relying on some private functions, too.

**Comment by lbittarello** [comment]
> Where does that convention come from? I though we mostly use underscores to denote functions which are private _for the users of glum_. I understand that these functions live in underscored modules [...], but why switch? I personally don't see a huge benefit in using names to denote whether functions are module-private or package-private.

At the moment, there's no pattern to which functions take a leading underscore. Some private functions don't have one (e.g., `is_pos_semidef`), some do. I tried to be more systematic. The advantage of this convention is that you can immediately tell whether other modules call a function or not when reading a file. We could also just use trailing underscores everywhere though.

> Why not simply `_glm_py`, and the current `_glm.py` could become `_glm_base.py`. That would be a bit more aligned with class names IMO.

It would, but I'm trying to minimise downstream breakage.

**Comment by stanmart** [comment]
> At the moment, there's no pattern to which functions take a leading underscore. [...]

In that case I'm fine with either, and agree that we should pick one and be consistent.

**Comment by stanmart** [comment]
Could [this](https://docs.python.org/3/library/doctest.html#doctest.NORMALIZE_WHITESPACE) solve the doctest issue?

**Comment by lbittarello** [comment]
> Could [this](https://docs.python.org/3/library/doctest.html#doctest.NORMALIZE_WHITESPACE) solve the doctest issue?

It didn't.

**Review by stanmart — APPROVED** [review]
LGTM but I imagine @jtilly and @MarcAntoineSchmidtQC have more to say
